src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+    /// which can be checked through the IsValidSummary, SummarizesOverConflictingResolution, and
+    /// SummarizesOverParameterConstriction flags.
+    /// </summary>
+    public class TypeResolutionSummary
+    {
+        // Static Members
+
+        /// <summary>
+        /// Checks if the given type parameter directly resolves to itself. If so, the
+        /// resolution may be overridden without conflict.
+        /// </summary>
+        private static bool IsSelfResolution(TypeParameterName typeParam, ResolvedType res)
+        {
+            return res.Resolution is ResolvedTypeKind.TypeParameter tp
+                && tp.Item.Origin.Equals(typeParam.Item1)
+                && tp.Item.TypeName.Equals(typeParam.Item2);
+        }
+
+        /// <summary>
+        /// Reverses the dependencies of type parameters resolving to other type parameters in the given
+        /// dictionary to create a lookup whose keys are type parameters and whose values are all the type
+        /// parameters that can be updated by knowing the resolution of the lookup's associated key.
+        /// </summary>
+        private static ILookup<TypeParameterName, TypeParameterName> GetReplaceable(TypeParameterResolutions.Builder typeParamResolutions)
+        {
+            return typeParamResolutions
+               .Select(kvp => (kvp.Key, GetTypeParameters.Apply(kvp.Value))) // Get any type parameters in the resolution type.
+               .SelectMany(tup => tup.Item2.Select(value => (tup.Key, value))) // For each type parameter found, match it to the dictionary key.
+               .ToLookup(// Reverse the keys and resulting type parameters to make the lookup.
+                   kvp => kvp.value,
+                   kvp => kvp.Key);
+        }
+
+        // Fields and Properties
+
+        /// <summary>
+        /// Array of all the type parameter resolution dictionaries that are summarized in this summary.
+        /// </summary>
+        public readonly ImmutableArray<TypeParameterResolutions> ResolutionsInSummary;
+
+        /// <summary>
+        /// The resulting resolution dictionary from combining all the input resolutions in
+        /// ResolutionsInSummary. Represents a summary of all the type parameters found and their
+        /// ultimate type resolutions.
+        /// </summary>
+        public TypeParameterResolutions CombinedTypeParameterResolutions { get; private set; }
+
+        /// <summary>
+        /// Flag for if there were any invalid scenarios encountered while creating the summary.
+        /// Invalid scenarios include a type parameter being assigned to multiple conflicting types
+        /// and a type parameter being assigned to a type referencing a different type parameter of
+        /// the same callable. Has value true if no invalid scenarios were encountered.
+        /// </summary>
+        public bool IsValidSummary { get => !this.SummarizesOverConflictingResolution && !this.SummarizesOverParameterConstriction; }
```

Comment:
I think if this was named `IsValid` it would be slightly better.

Reply:
Done.